import/index_java/src/org/vufind/index/DoiTools.java

```diff
@@ -44,7 +64,36 @@ public Set<String> getDoiFromUrl(final Record record, String fieldSpec, String b
         for (String current: input) {
             // If the base URL is found in the string, crop it off for our DOI!
             if (current.startsWith(baseUrl)) {
-                result.add(current.substring(baseUrl.length()));
+                result.add(decodeDoi(current.substring(baseUrl.length())));
+            }
+        }
+
+        // If we found no matches, return null; otherwise, return our results:
+        return result.isEmpty() ? null : result;
+    }
+
+    /**
+     * Extract DOIs from URLs matching the specified regular expression
+     * @param record MARC record
+     * @param fieldSpec taglist for URL fields
+     * @param regEx Regular expression for matching URLs
+     * @param groupIndex Index of the grouped expression in regEx containing the DOI to extract (passed as a String instead of an Integer due to SolrMarc limitations)
+     * @return Set of DOIs
+     */
+    public Set<String> getDoisFromUrlWithRegEx(final Record record, String fieldSpec, String regEx, String groupIndex) {
+        // Build the regular expression:
+        Pattern pattern = Pattern.compile(regEx);
```

Comment:
This method might be called millions of times, and it will always compile the same pattern over and over again. Can we use some kind of cache for compiled patterns?

Or maybe we could use something like the CachedPattern class in this article:
https://stackoverflow.com/questions/13420321/does-pattern-compile-cache

(We should also keep in mind that the cache should be thread-safe)

Reply:
Thanks, @mtrojan-ub -- what do you think of [e24ee32](https://github.com/vufind-org/vufind/pull/2473/commits/e24ee32bc84a7e2660fb9e55846121e259557e2f)?

Reply:
Good point, @mtrojan-ub, that's a lot nicer -- see [1d867dc](https://github.com/vufind-org/vufind/pull/2473/commits/1d867dce856f4e194835ceb8b93b8f1428059a37).

Reply:
(Also accidentally introduced unwanted whitespace into the comment in that commit -- fixed in a subsequent commit; sorry about that!)

Reply:
Looks fine now, thanks!

Reply:
Thanks again for the very helpful input, @mtrojan-ub!